### PR TITLE
Adrienne / Integrate Hydra authentication

### PR DESCRIPTION
**Pull request by adrienne-deriv**
Changes:

- Setup `@deriv-com/analytics` to integrate feature toggling from Growthbook for enabling/disabling Hydra authentication flow
- Setup `@deriv-com/auth-client` to integrate the new Hydra authentication flow
- Refactored the Deriv iframe `localstorage-sync` to be enabled/disabled based on the feature flag from Growthbook

## Type of change

-   [ ] Bug fix
-   [x] New feature
-   [ ] Update feature
-   [ ] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [ ] Dependency update
-   [ ] Documentation update
-   [ ] Release

### CODE REVIEW

File: src/javascript/app/pages/deriv_iframe.jsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        style={{ display: 'none', visibility: 'hidden' }}
+        sandbox='allow-same-origin allow-scripts'
+    />
+);
+
+export const init = () => {
+    const isAuthEnabled = isOAuth2Enabled();
+
+    if (!isAuthEnabled) ReactDOM.render(<DerivIFrame />, document.getElementById('deriv_iframe'));
+};
```
Comment (adrienne-deriv) [inline src/javascript/app/pages/deriv_iframe.jsx:17]: I moved the `DerivIframe` component to the `javascript` folder so that it can be conditionally rendered to be included/excluded based on the feature flag returned by `isOauth2Enabled`.
If the Hydra authentication is enabled, we will not be using the Deriv iframe to sync login states anymore

File: src/javascript/app/base/header.js [hunk 2]
```diff
@@ -624,8 +626,10 @@ const Header = (() => {
         Login.redirectToLogin();
     };
 
-    const logoutOnClick = () => {
-        Client.sendLogoutRequest();
+    const logoutOnClick = async () => {
+        const onLogoutWithOauth = await AuthClient.getLogoutHandler(Client.sendLogoutRequest);
+
```
Comment (adrienne-deriv) [inline src/javascript/app/base/header.js:635]: This will wrap the logout call `Client.sendLogoutRequest` with our own logout iframe, which is to inform Hydra that the user is logging out and the session should be cleared on Hydra's side. Once this is done, it will call the passed-in logout handler `Client.sendLogoutRequest`.
If Hydra authentication is not enabled, the logout handler `Client.sendLogoutRequest` will just be called instead.

File: src/javascript/_common/base/__tests__/client_base.js [hunk 3]
```diff
@@ -1,3 +1,10 @@
+
+const Mock                         = require('mock-require');
+Mock('../../auth', {
+    isOAuth2Enabled: function() {
+        return false
+    }
+});
```
Comment (adrienne-deriv) [inline src/javascript/_common/base/__tests__/client_base.js:7]: This is to resolve test case issues where `@deriv-com/analytics` is using Growthbook module which uses MutationObserver, in which Mocha doesn't recognise and is unable to mock it
Screenshot 2024-09-26 at 11 53 04 AM

File: .github/workflows/release_production.yml [hunk 4]
```diff
@@ -22,6 +23,9 @@ jobs:
         uses: "./.github/actions/build"
         with:
           target: production
+          RUDDERSTACK_KEY:  ${{ vars.RUDDERSTACK_KEY }}
+          GROWTHBOOK_DECRYPTION_KEY: ${{ secrets.GROWTHBOOK_DECRYPTION_KEY }}
```
Comment (amam-deriv) [inline .github/workflows/release_production.yml]: [Q] why is this on secrets but client key is on vars?

File: src/javascript/app/pages/deriv_iframe.jsx (hunk 1 above)
Comment (henry-deriv) [reply src/javascript/app/pages/deriv_iframe.jsx:17]: so we want to disable syncing in certain scenarios?

Comment (adrienne-deriv) [reply src/javascript/app/pages/deriv_iframe.jsx:17]: We only disable it when the Hydra authentication feature flag is enabled, which we turn on/off from Growthbook

Comment (adrienne-deriv) [reply src/javascript/app/pages/deriv_iframe.jsx:17]: Because one of our aim for this feature is to remove the use of Deriv iframe sync entirely and migrate these syncing logic into the Hydra service, which is on BE's side

**Review by shayan-deriv — APPROVED** [review]

**Review by ahmadtaimoor-deriv — DISMISSED** [review]

**Review by thisyahlen-deriv — APPROVED** [review]

File: src/javascript/_common/auth.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    const oauthUrl = appId && serverUrl ? `https://${serverUrl}/oauth2/sessions/logout` : DEFAULT_OAUTH_LOGOUT_URL;
+
+    return oauthUrl;
+};
+
+export const getOAuthOrigin = () => {
+    const { appId, serverUrl } = getServerInfo();
+
+    const oauthUrl = appId && serverUrl ? `https://${serverUrl}` : DEFAULT_OAUTH_ORIGIN_URL;
```
Comment (amam-deriv) [inline src/javascript/_common/auth.js:59]: [Q] whats appId check used here for?

File: src/javascript/_common/auth.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    if (!isAuthEnabled) {
+        return onWSLogoutAndRedirect;
+    }
+
+    const onMessage = async event => {
+        const allowedOrigin = getOAuthOrigin();
+        if (allowedOrigin === event.origin) {
+            if (event.data === 'logout_complete') {
+                await onWSLogoutAndRedirect();
```
Comment (amam-deriv) [inline src/javascript/_common/auth.js]: [R] please wrap await function in a try catch block

File: src/javascript/_common/auth.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        let iframe = document.getElementById('logout-iframe');
+        if (!iframe) {
+            iframe = document.createElement('iframe');
+            iframe.id = 'logout-iframe';
+            iframe.style.display = 'none';
+            document.body.appendChild(iframe);
+
+            setTimeout(() => {
+                onWSLogoutAndRedirect();
+            }, 10000);
```
Comment (amam-deriv) [inline src/javascript/_common/auth.js]: [M] static data can be put in a constant variable on top of the file

File: src/javascript/app/base/header.js (hunk 2 above)
Comment (amam-deriv) [reply src/javascript/app/base/header.js:635]: this explanation is great, though you can also add it as comment too

File: src/javascript/app/hooks/growthbook.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import { Analytics } from '@deriv-com/analytics';
+
+export const useIsGrowthbookIsLoaded = () => {
+    const [isGBLoaded, setIsGBLoaded] = useState(false);
+
+    useEffect(() => {
+        let checksCounter = 0;
+        const analyticsInterval = setInterval(() => {
+            // Check if the analytics instance is available for 10 seconds before setting the feature flag value
+            if (checksCounter > 20) {
```
Comment (amam-deriv) [inline src/javascript/app/hooks/growthbook.js]: not sure why is this required, but i see that this is carried over implementation.

File: src/javascript/app/hooks/growthbook.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                // If the analytics instance is not available after 10 seconds, clear the interval
+                clearInterval(analyticsInterval);
+                return;
+            }
+            checksCounter += 1;
+            if (Analytics?.getInstances()?.ab) {
+                setIsGBLoaded(true);
+                clearInterval(analyticsInterval);
+            }
+        }, 500);
```
Comment (amam-deriv) [inline src/javascript/app/hooks/growthbook.js]: [M] this can become 1000 and checksCounter > 10?

**Review by amam-deriv — CHANGES_REQUESTED** [review]

File: .github/workflows/release_production.yml (hunk 4 above)
Comment (adrienne-deriv) [reply .github/workflows/release_production.yml]: The Rudderstack and Growthbook client keys are public, you can see them on the network requests and `window._growthbook` instances, but only the decryption keys are private

File: src/javascript/_common/auth.js (hunk 5 above)
Comment (adrienne-deriv) [reply src/javascript/_common/auth.js:59]: Its just an extra check to ensure we have the app ID and server URL before returning the origin url

File: src/javascript/app/hooks/growthbook.js (hunk 8 above)
Comment (adrienne-deriv) [reply src/javascript/app/hooks/growthbook.js]: No this hook file is no longer needed, will remove it for now 😅

**Review by thisyahlen-deriv — APPROVED** [review]

File: src/javascript/_common/auth.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    // we will use the red server with app_id=62019 for the staging-p2p.deriv.com for now
+    // to fix the login issue
+    if (origin === URLConstants.derivP2pStaging && (!existingAppId || !existingServerUrl)) {
+        LocalStorageUtils.setValue(LocalStorageConstants.configServerURL, SocketURL[origin]);
+        LocalStorageUtils.setValue(LocalStorageConstants.configAppId, `${AppIDConstants.domainAppId[hostname]}`);
+    }
+
+    const serverUrl = LocalStorageUtils.getValue(LocalStorageConstants.configServerURL) || localStorage.getItem('config.server_url') || 'oauth.deriv.com';
+
+    const appId = LocalStorageUtils.getValue(LocalStorageConstants.configAppId);
```
Comment (amam-deriv) [inline src/javascript/_common/auth.js]: [Q] is the value always available that we don't need a fallback?

**Review by amam-deriv — APPROVED** [review]
LGTM!

**Review by thisyahlen-deriv — APPROVED** [review]

**Review by ahmadtaimoor-deriv — APPROVED** [review]
